PMM-10159 Mask sensitive data for alertmanager

gen1us2k opened the pull request.

gen1us2k committed:
PMM-10159 Mask sensitive data

gen1us2k committed:
PMM-10159 Specified masked struct fields

gen1us2k committed:
PMM-10159 Fixed build

gen1us2k committed:
PMM-10159 Added license header

gen1us2k committed:
PMM-10159 Small refactoring

gen1us2k committed:
PMM-10159 Reduce cyclomatic complexity

gen1us2k committed:
PMM-10159 More tests

gen1us2k committed:
PMM-10159 Added support for struct

gen1us2k committed:
PMM-10159 make format

gen1us2k committed:
PMM-10159 Added more tests

gen1us2k committed:
PMM-10159 make linter happy

gen1us2k committed:
PMM-10159 make linter happy

gen1us2k committed:
PMM-10159 make linter happy

gen1us2k committed:
PMM-10159 drop switch case

BupycHuk commented on alertmanager/notifiers.go:
```diff
@@ -167,10 +167,10 @@ type WechatConfig struct {
+	CorpID      string `yaml:"corp_id,omitempty" masked:"true"`
 	Message     string `yaml:"message,omitempty"`
-	APIURL      string `yaml:"api_url,omitempty"`
+	APIURL      string `yaml:"api_url,omitempty" masked:"true"`
```
let's not mask API URLs, I think they might be helpful

BupycHuk replied:
here and in other places

artemgavrilov commented on alertmanager/mask.go:
```diff
@@ -0,0 +1,72 @@
+
+	for i := 0; i < val.NumField(); i++ {
+		f := val.Field(i)
+		if f.Kind() == reflect.Ptr {
+			if f.IsNil() {
+				continue
+			}
+			MaskSensitiveData(f.Interface())
+		}
+		if f.Kind() == reflect.Struct {
+			MaskSensitiveData(f.Addr().Interface())
+		}
+		if f.Kind() == reflect.Slice {
+			for j := 0; j < f.Len(); j++ {
+				MaskSensitiveData(f.Index(j).Interface())
+			}
+		}
+		if f.Kind() == reflect.String {
+			// masked struct tag must be equal to true to mask values
+			processString(val.Type().Field(i).Tag.Get("masked"), f)
+		}
```
Switch will make it easier to read

gen1us2k replied:
I agree, but linter throws errors that some types are unhandled

artemgavrilov commented on alertmanager/mask.go:
```diff
@@ -0,0 +1,72 @@
+const maskedValue = "xxxxxxxx"
+
+// MaskSensitiveData loops over Config struct and masks sensitive data.
+func MaskSensitiveData(c interface{}) {
```
Let's move it to the project root. We may want to use it also for prometheus configs as well.

artemgavrilov commented on alertmanager/mask.go:
```diff
@@ -0,0 +1,72 @@
+	b, err := strconv.ParseBool(s)
+	if err != nil {
+		return false
+	}
```
Minor: I think it can be simplified just like this
```suggestion
			// masked struct tag must be equal to true to mask values
			if val.Type().Field(i).Tag.Get("masked") == "true" && f.CanSet() && f.String() != "" {
				f.SetString(maskedValue)
			}
		}
	}
}

```

artemgavrilov commented on alertmanager/mask.go:
```diff
@@ -0,0 +1,72 @@
+const maskedValue = "xxxxxxxx"
+
+// MaskSensitiveData loops over Config struct and masks sensitive data.
+func MaskSensitiveData(c interface{}) {
```
Too bad that this function modifies original object. Maybe we can create method for `Config`, something like:
```
func (c *Config) Mask *Config {
    nc := c.Copy()
    MaskSensitiveData(nc)
    return nc
}
```

In the end we don't need generic solution. What do you think?

gen1us2k replied:
Yep. I had the same concern about it. Moving the code to the root directory and implementing wrappers for `alertmanager.Config` and `promconfig.Config` would be best.

gen1us2k committed:
Update alertmanager/mask.go

Co-authored-by: Artem Gavrilov <artem.gavrilov@percona.com>

gen1us2k committed:
PMM-10159 Implement Mask method for alertmanager.Config

A small refactoring of the old code

* Moved MaskSensitiveData to the root of the project
* Implemented alertmanager.Config.Mask() method that uses MaskSensitiveData
* Use switch instead of multiple ifs

gen1us2k committed:
PMM-10159 go fmt

gen1us2k committed:
PMM-10159 nolint switch case

gen1us2k committed:
PMM-10159 package comment added

gen1us2k committed:
PMM-10159 handle type cast

gen1us2k committed:
PMM-10159 make linter happy

gen1us2k committed:
PMM-10159 tidy go.mod

BupycHuk commented on mask.go:
```diff
@@ -0,0 +1,53 @@
+const maskedValue = "xxxxxxxx"
+
+// MaskSensitiveData loops over Config struct and masks sensitive data.
+func MaskSensitiveData(c interface{}) { //nolint:cyclop
```
I wonder if we can use generics here to not use reference and return a value?
```suggestion
func MaskSensitiveData[T interface{}](c T) T { //nolint:cyclop
	nc, ok := deepcopy.Copy(c).(T)
	if !ok {
		return nil, errors.New("failed to copy config")
	}
        val := reflect.ValueOf(nc)
        ...
        return nc
```

artemgavrilov replied:
Yep, it will be nice

BupycHuk replied:
@gen1us2k looks like we keep updating fields of an existing object. Is it possible to make it mutable to not update the existing object?

gen1us2k replied:
Yes. But what problem do we want to solve? I think that probability of misusing Mask method is too low :)

artemgavrilov commented on alertmanager/config.go:
```diff
@@ -16,8 +16,19 @@
+import (
+	"errors"
+
+	"github.com/mohae/deepcopy"
```
`promconfig` is zero-dependency lib, we highlighting this in readme. Existing deps used only in tests, all regular lib files has no external dependencies. Also it seems that `deepcopy` a little bit abandoned.

gen1us2k committed:
Update mask.go

Co-authored-by: Nurlan Moldomurov <nurlan.moldomurov@percona.com>

gen1us2k commented:
I found that this feature is already implemented in the upstream project. 
I think that it's time to upgrade the code of this package
E.g.
https://github.com/prometheus/alertmanager/blob/main/config/notifiers.go#L169
https://github.com/prometheus/common/blob/main/config/config.go#L27

artemgavrilov commented:
> I found that this feature is already implemented in the upstream project. I think that it's time to upgrade the code of this package E.g. https://github.com/prometheus/alertmanager/blob/main/config/notifiers.go#L169 https://github.com/prometheus/common/blob/main/config/config.go#L27

Awesome! Let's do it!

gen1us2k committed:
Revert "Update mask.go"

This reverts commit 3da7e2c2c8f4e77df14f77657035edfcf71b09cc.

gen1us2k committed:
PMM-10159 Added secret data type

gen1us2k committed:
PMM-10159 Moved to Secret data type

gen1us2k committed:
PMM-10159 Drop dead code

gen1us2k committed:
PMM-10159 Drop dead deps

gen1us2k commented:
It's still a draft. Going to implement wrapper around the package to load files without any modifications

gen1us2k commented:
I updated configs from the upstream. Added a secret data type and fixed tests for them. However, the changes are backward incompatible. I think that I implement `Mask` method tomorrow that looks for the secret data type and overrides value. Without this PMM will override values with `<secret>` which breaks everything configured by users.

gen1us2k committed:
PMM-10159 Safe masking for Secret datatype

gen1us2k committed:
PMM-10159 Parallel test support for secret datatype

gen1us2k committed:
PMM-10159 code format

gen1us2k committed:
PMM-10159 missed docstrings

gen1us2k committed:
PMM-10159 added dots

gen1us2k committed:
PMM-10159 Make linter happy

gen1us2k committed:
PMM-10159 make linter happy

gen1us2k committed:
PMM-10159 Make linter happy

gen1us2k committed:
PMM-10159 Drop dead code

gen1us2k commented on secret.go:
```diff
@@ -0,0 +1,22 @@
+type Secret string
+
+// MarshalYAML implements the yaml.Marshaler interface for Secrets.
+func (s Secret) MarshalYAML() (interface{}, error) {
```
@BupycHuk @artemgavrilov I changed logic here a little bit and remove logic that masks values by default. In that case, these changes will break everything for PMM at least because it does config management there.

I restored the old code and changed it a little bit. In that case, we have yaml.Marshal and yaml.Unmarshal that does not change masked values and config.Mask() to mask everything.

I thought about copying the struct and modifying values only on the copied struct, but it seems a little bit redundant because of the following reasons

1.  People usually work with configs by doing Unmarshal, changing values, and marshal again. We don't have any syntax sugar as upstream config have
2. Mask method modifies values of unmarshaled struct in that case it can lead to some misuse in the future. Right now we will use it only for masking sensitive data on /logs.zip endpoint. I think we can add copying of struct a little bit later
What do you think?

artemgavrilov replied:
If we don't have custom `Marshal/Unmarshal` functions then it seems it has no sense to have custom `Secret` type for that. Changing field type is breaking change, we should release lib like `v2` and user should update his sources. If we plan to use `Mask` function that replaces sensitive data then tag `secret:true` looks better. It will be a minor release and will not affect anyone.

artemgavrilov commented on mask.go:
```diff
@@ -0,0 +1,51 @@
+
+import "reflect"
+
+// MaskSecret loops over Config struct and masks values for Secret datatype.
```
```suggestion
// MaskSecret loops over struct and masks values for Secret datatype.
```

gen1us2k committed:
Update mask.go

Co-authored-by: Artem Gavrilov <artem.gavrilov@percona.com>

gen1us2k committed:
PMM-10159 Drop Secret datatype

gen1us2k committed:
PMM-10159 Copy config before masking

gen1us2k committed:
PMM-10159 formatted

gen1us2k committed:
PMM-10159 make linter happy

gen1us2k commented:
@artemgavrilov We discussed with @BupycHuk about licensing and copying dependencies, so it should look fine. I copied and simplified deepcopy and added it to promconfig package. It looks good to merge I think.

artemgavrilov reviewed (approved):
👍

gen1us2k committed:
Merge branch 'main' into PMM-10159-mask-sensitive-data

artemgavrilov commented:
After merge we should not forget to add new version tag
